### PR TITLE
fix: lam rollout

### LINKED ISSUE #212: LAM boundary rollout incorrectly implemented
### What happened?

During rollout with a lam model, after a first 6h forecast step, the data outside of the cutout region needs to be replaced with data from the training dataset (rather than remain a forecast, as inside the cutout). This is currently incorrectly implemented in anemoi-training due to a misinterpretation of the workings of `torch.Tensor.masked_scatter`.  

### What are the steps to reproduce the bug?

I will present here a toy example that illustrates the issue. I have verified myself that this toy example correctly reproduces what is going wrong in actual anemoi training runs by saving some of the relevant tensors to disk, and checking that `boundary_update_training` as defined below reproduces the boundary replacement as performed by anemoi-training.

Let me first point to the relevant parts of the actual anemoi code. Boundary rollout as performed by anemoi training can be found [here](https://github.com/ecmwf/anemoi-core/blob/833e0e7ea905dcd8f38800ef3f49747282ddc81f/training/src/anemoi/training/utils/masks.py#L60). Boundary rollout as performed in anemoi-inference can be found [here](https://github.com/ecmwf/anemoi-inference/blob/c645fb219ed40ea60389affafe0f4f6c3b799ee9/src/anemoi/inference/runner.py#L759).

The difference between the two methods, and why the method used in training is erroneous can be seen by running the following simple python script:
```python
import numpy as np
import torch

def boundary_update_inference(input, fill_value, spatial_mask):
    updated_input = input.copy()
    updated_input[spatial_mask] = fill_value[spatial_mask]
    return updated_input

def boundary_update_training(input, fill_value, spatial_mask):
    updated_input = input.copy()
    x = torch.tensor(updated_input)
    fill_value = torch.tensor(fill_value)
    spatial_mask = torch.tensor(spatial_mask)
    x = x.masked_scatter(spatial_mask, fill_value)
    return x.numpy()

test = np.array([-1 for i in range(10)])
mask = np.array([ i > 3 for i in range(10)])
fill = np.array([ i for i in range(10)])

print("consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')")
print(f"test: {test}")
print(f"fill: {fill}")
print(f"boundary mask: {mask}")
print(f"replacement of test with fill on 'boundary' as in inference [correct]: {boundary_update_inference(toy_test,toy_fill,toy_mask)}")
print(f"replacement of test with fill on 'boundary' as in training [incorrect]: {boundary_update_training(toy_test,toy_fill,toy_mask)}")
```
The output of this script is
```
consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')
test: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1]
fill: [0 1 2 3 4 5 6 7 8 9]
boundary mask: [False False False False  True  True  True  True  True  True]
replacement of test with fill on 'boundary' as in inference [correct]: [-1 -1 -1 -1  4  5  6  7  8  9]
replacement of test with fill on 'boundary' as in training [incorrect]: [-1 -1 -1 -1  0  1  2  3  4  5]
```
One sees that `boundary_update_training` actually replaces the value of test at the first boundary gridpoint (index 4 in this example) with the value of fill at the first gridpoint (index 0 in this example) which is actually inside the cutout region! On the contrary  `boundary_update_inference` replaces the value of test at the first boundary gridpoint with the value of fill at the same gridpoint, which is what boundary rollout is supposed to do.

As mentioned above I have checked that the exact same problem can be observed when saving the tensors produced during training to disk and analysing them.

### Version

latest version

### Platform (OS and architecture)

irrelevant

### Relevant log output

```shell
You wont see this by looking at training logs.
```

### Accompanying data

_No response_

### Organisation

RMI

### PR DESCRIPTION
**Pull request by dietervdb-meteo**
## Description

`torch.Tensor.masked_scatter` was used incorrectly, this is easily fixed by first using the mask to select that part of the fill values that should actually be used by `masked_scatter` for filling. 

Update: in the last version of the code we decided to forego the use of `masked_scatter` completely. The idea behind the fix remains the same though, namely first selecting that part of `fill_value` that we actually want to use for filling.

## Type of Change

-   [x] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] Documentation update

## Issue Number

Closes #212. For completeness one can find an update of the toy python script provided in that issue below. The function `boundary_update_training_fixed` is a toy version of the fix implemented in this PR. Be aware that the simple code below is only for illustrating the main idea. For the actual implementation in anemoi see the *Files changed* section of this PR, that implementation is an extension of the simple idea below to multi-dimensional tensors. 

```python
import numpy as np
import torch

def boundary_update_inference(input, fill_value, spatial_mask):
    updated_input = input.copy()
    updated_input[spatial_mask] = fill_value[spatial_mask]
    return updated_input

def boundary_update_training(input, fill_value, spatial_mask):
    updated_input = input.copy()
    x = torch.tensor(updated_input)
    fill_value = torch.tensor(fill_value)
    spatial_mask = torch.tensor(spatial_mask)
    x = x.masked_scatter(spatial_mask, fill_value)
    return x.numpy()

def boundary_update_training_fixed(input, fill_value, spatial_mask):
    updated_input = input.copy()
    x = torch.tensor(updated_input)
    fill_value = torch.tensor(fill_value)
    spatial_mask = torch.tensor(spatial_mask)
    x = x.masked_scatter(spatial_mask, fill_value[spatial_mask])
    return x.numpy()

test = np.array([-1 for i in range(10)])
mask = np.array([ i > 3 for i in range(10)])
fill = np.array([ i for i in range(10)])

print("consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')")
print(f"test: {test}")
print(f"fill: {fill}")
print(f"boundary mask: {mask}")
print(f"replacement of test with fill on 'boundary' as in inference [correct]: {boundary_update_inference(toy_test,toy_fill,toy_mask)}")
print(f"replacement of test with fill on 'boundary' as in training [incorrect]: {boundary_update_training(toy_test,toy_fill,toy_mask)}")
print("consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')")
print(f"test: {toy_test}")
print(f"fill: {toy_fill}")
print(f"boundary mask: {toy_mask}")
print(f"replacement of test with fill on 'boundary' as in inference [correct]: {boundary_update_inference(toy_test,toy_fill,toy_mask)}")
print(f"replacement of test with fill on 'boundary' as in training [incorrect]: {boundary_update_training(toy_test,toy_fill,toy_mask)}")
print(f"replacement of test with fill on 'boundary' as in training_fixed [correct]: {boundary_update_training_fixed(toy_test,toy_fill,toy_mask)}")
```

which produces as output
```
consider the following toy example (i <= 3 is 'cutout', i > 3 is 'boundary')
test: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1]
fill: [0 1 2 3 4 5 6 7 8 9]
boundary mask: [False False False False  True  True  True  True  True  True]
replacement of test with fill on 'boundary' as in inference [correct]: [-1 -1 -1 -1  4  5  6  7  8  9]
replacement of test with fill on 'boundary' as in training [incorrect]: [-1 -1 -1 -1  0  1  2  3  4  5]
replacement of test with fill on 'boundary' as in training_fixed [correct]: [-1 -1 -1 -1  4  5  6  7  8  9]
```

## Code Compatibility

-   [x] I have performed a self-review of my code

### Code Performance and Testing

-   [x] I have done training tests with the relevant tensors saved to disk and checked that indeed the boundary update is done correctly, which shows my fix is effective
-   [ ] I ran the [complete Pytest test](https://anemoi.readthedocs.io/projects/training/en/latest/dev/testing.html) suite locally, and they pass
-   [x] I have tested the changes on a single GPU
-   [ ] I have tested the changes on multiple GPUs / multi-node setups
-   [ ] I have run the Benchmark Profiler against the old version of the code
-   [ ] If the new feature introduces modifications at the config level, I have made sure to update Pydantic Schemas and default configs accordingly

### Dependencies

-   [x] I have ensured that the code is still pip-installable after the changes and runs
-   [ ] I have tested that new dependencies themselves are pip-installable.
-   [ ] I have not introduced new dependencies in the inference portion of the pipeline

### Documentation

-   [ ] My code follows the style guidelines of this project
-   [x] I have updated the documentation and docstrings to reflect the changes
-   [ ] I have added comments to my code, particularly in hard-to-understand areas

### CODE REVIEW

**Comment by dietervdb-meteo** [comment]
I checked that the fix works by training a model, performing rollout and running verification after inference. One now explicitly sees that rollout training improves the performance at later lead times, as it should. Contrary to before the fix.

**Review by JPXKQX — APPROVED** [review]

**Review by anaprietonem — APPROVED** [review]
